CB-10612 Stack service updates for in-place enablement of load balanacers

hreeve-cloudera opened the pull request:
Fleshes out the Stack service update_load_balancers logic. This API call is internal only, and is
intended to be called by the Environment service with an environment load balancer update is
triggered. A new flow was added to update an existing environment with load balancers. The flow
will in order:
  1. Create the load balancer database entities and save them.
  2. Create the cloud load balancer resources.
  3. Collect the load balancer cloud metadata and persist it to the database.
  4. Register the load balancers' public DNS entries with Route 53.
  5. Register the load balancers' private DNS entries with FreeIPA.
  6. Update the CM and Hue services to have the appropriate load balancer DNS settings.
  7. Restart the CM server to pick up the updated setting.

Some logic has been refactored so that the load balancer related logic can be executed independently
of the rest of the stack creation logic. In summary:
* The ModelContext creation logic has been moved out of AwsLaunchService and into a new AwsModelService.
* The LB creation logic has been moved out of AwsLaunchService and into a new AwsLoadBalancerService.
* The LB metadata collection logic has been moved out of Stack creation CollectLoadBalancerMetadataHandler action and into a new LoadBalancerMetadataService.
* Some previously private methods were made public so they can be called as part of the new update flow.

Tested with unit tests and by running the new API commands against a stack without a load balancer
in place, and validating the database and cloud resources afterwards

Currently based on branch CB-10612-b. Will be rebased on master when that change is merged in.

doktoric commented:
Please make sure that the PR check are green and then we will do the review

enis commented:
Just pushed a rebased version that also addresses some of the build issues like spot bugs, etc until @hreeve-cloudera gets the power back. I think this may still be failng the IT, but rest of it should hopefully be green.

doktoric commented on cloud-api/src/main/java/com/sequenceiq/cloudbreak/cloud/model/CloudLoadBalancerMetadata.java:
```diff
@@ -12,11 +12,14 @@
 
     private final String ip;
 
-    public CloudLoadBalancerMetadata(LoadBalancerType type, String cloudDns, String hostedZoneId, String ip) {
```
would nice to see a builder pattern here

doktoric commented on cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureResourceConnector.java:
```diff
@@ -316,4 +316,11 @@ public String getStackTemplate() {
+    public List<CloudResourceStatus> updateLoadBalancers(AuthenticatedContext authenticatedContext, CloudStack stack,
+            PersistenceNotifier persistenceNotifier) {
+        // no-op
+        return List.of();
```
should we throw an expception here ?

doktoric commented on cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpResourceConnector.java:
```diff
@@ -138,4 +139,11 @@ public String getDBStackTemplate() throws TemplatingNotSupportedException {
+    public List<CloudResourceStatus> updateLoadBalancers(AuthenticatedContext authenticatedContext, CloudStack stack,
+            PersistenceNotifier persistenceNotifier) {
+        // no-op
+        return List.of();
```
maybe a not supported exception ?

doktoric commented on cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/service/LoadBalancerMetadataService.java:
```diff
@@ -0,0 +1,36 @@
+        CloudConnector<Object> connector = cloudPlatformConnectors.get(cloudContext.getPlatformVariant());
+        AuthenticatedContext ac = connector.authentication().authenticate(cloudContext, cloudCredential);
+        List<CloudLoadBalancerMetadata> loadBalancerStatuses = connector.metadata().collectLoadBalancer(ac, types);
+        LOGGER.debug("Load balancer metadata collection successfully finished");
```
can we show the result here ?

doktoric commented on cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnResourceConnector.java:
```diff
@@ -281,6 +281,13 @@ public String getDBStackTemplate() throws TemplatingNotSupportedException {
+    public List<CloudResourceStatus> updateLoadBalancers(AuthenticatedContext authenticatedContext, CloudStack stack,
+            PersistenceNotifier persistenceNotifier) {
+        // no-op
+        return List.of();
```
exception here ?

doktoric commented on cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerConfigService.java:
```diff
@@ -101,4 +105,38 @@ private void modifyKnoxAutorestart(ApiClient client, String clusterName, boolean
         }
     }
+
+    public void modifyHueKnoxProxyHosts(ApiClient client, String clusterName, String knoxProxyhosts) {
```
seems this is really hue specific? can we do a little bit more generic solution here? I guess in the future there will be more

doktoric commented on core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/common/Status.java:
```diff
@@ -50,7 +50,11 @@
     EXTERNAL_DATABASE_STOP_FINISHED(StatusKind.PROGRESS),
-    EXTERNAL_DATABASE_STOP_FAILED(StatusKind.FINAL);
+    EXTERNAL_DATABASE_STOP_FAILED(StatusKind.FINAL),
+    LOAD_BALANCER_UPDATE_IN_PROGRESS(StatusKind.PROGRESS),
```
this is probably requires UI changes

hreeve-cloudera replied:
Opened https://jira.cloudera.com/browse/CB-11308 for Santiago to work on the UI changes.

doktoric commented on core/src/main/java/com/sequenceiq/cloudbreak/controller/v4/StackV4Controller.java:
```diff
@@ -365,6 +365,12 @@ public FlowIdentifier renewCertificate(Long workspaceId, String name, @Initiator
 
+    @Override
+    @CheckPermissionByAccount(action = AuthorizationResourceAction.POWERUSER_ONLY)
+    public FlowIdentifier updateLoadBalancers(Long workspaceId, String name, @AccountId String accountId) {
```
do we need a crn based endpoint as well ? cc @lnardai

doktoric commented on core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/StackV4RequestToStackConverter.java:
```diff
@@ -178,7 +178,7 @@ public Stack convert(StackV4Request source) {
         determineServiceTypeTag(stack, source.getTags());
         determineServiceFeatureTag(stack, source.getTags());
-        stack.setLoadBalancers(loadBalancerConfigService.createLoadBalancers(stack, environment));
+        stack.setLoadBalancers(loadBalancerConfigService.createLoadBalancers(stack, environment, stack.getInstanceGroups()));
```
we are passing the stack and the stack.getInstanceGroups() why is that ?

doktoric commented on core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/config/generator/OfflineStateGenerator.java:
```diff
@@ -327,6 +327,11 @@ public ConfigurableListableBeanFactory getBeanFactory() {
         }
 
+        @Override
+        public List<Stack> findByEnvironmentCrnStackTypeAndWorkspaceId(String environmentCrn, StackType type, Long workspaceId) {
```
is this correct? or just a missing implementation?

doktoric commented on environment/src/main/java/com/sequenceiq/environment/environment/service/LoadBalancerPollerService.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    private Set<String> getAttachedDatalakeClusters(String environmentName) {
+        LOGGER.debug("Getting SDX clusters for environment: '{}'", environmentName);
```
In the logs there are Datahubs/Data Hubs/Datalakes/Data Lakes/SDX can we add this terms consistent ?

doktoric commented on core/src/main/resources/application.yml:
```diff
@@ -654,8 +654,8 @@ clusterProxy:
       attempt: 50
       intervall: 10
-    base.domain.name: cloudera.com
+    base.domain.name: dev.cldr.work
```
do we sure that we would like to change it ? cc @keyki

doktoric commented on core/src/main/java/com/sequenceiq/cloudbreak/repository/InstanceMetaDataRepository.java:
```diff
@@ -121,4 +121,7 @@
     int updateStatusIfNotTerminated(@Param("id") Long id, @Param("newInstanceStatus") InstanceStatus newInstanceStatus,
             @Param("newStatusReason") String newStatusReason);
 
+    @EntityGraph(value = "InstanceMetaData.instanceGroup", type = EntityGraph.EntityGraphType.LOAD)
```
I am not sure that we would like to do this fetching cc @akanto

hreeve-cloudera replied:
Can you elaborate on why we shouldn't do this fetching?

I added this because the `GatewayPublicEndpointManagementService.renewCertificates` call ultimately gets the stack endpoint by calling `Stack.getPrimaryGatewayInstance()`, which in turn accesses `InstanceGroup.instanceMetadata`. I was running into errors because that field wasn't being loaded due to lazy loading, so I added this call to get it and set it manually.

I think instead of getting all instance groups and all instance metadata, I can call `InstanceGroupRepository.getPrimaryGatewayInstanceGroupByStackId` and `InstanceMetaDataRepository.getPrimaryGatewayInstanceMetadata` to populate just the objects I really need. Would that work?

hreeve-cloudera replied:
I went ahead and made the change I mentioned, and removed this new method.

doktoric commented on cluster-api/src/main/java/com/sequenceiq/cloudbreak/cluster/api/ClusterModificationService.java:
```diff
@@ -31,6 +31,8 @@
 
     Map<String, String> gatherInstalledParcels(String stackName);
 
+    void updateHueKnoxProxyhosts(String knoxProxyhosts) throws Exception;
```
this is not really  generic. I would prefer updateServiceConfiguration with possible ways to add more service

hreeve-cloudera commented:
The rebase last week accidentally pulled in the changes for 9963 into this change. My most recent push separates that out again.

hreeve-cloudera commented:
@doktoric I rebased this PR after merging in https://github.com/hortonworks/cloudbreak/pull/9963, so if you want to test it you don't need to merge anything together. You should be able to test this branch directly now.

biharitomi commented on core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackService.java:
```diff
@@ -393,6 +393,10 @@ public StackV4Response getByAmbariAddress(String ambariAddress) {
         return stackRepository.findByEnvironmentCrnAndStackType(environmentCrn, stackType);
     }
 
+    public List<Stack> getStackByEnvironmentCrnAndStackType(String environmentCrn, StackType stackType, Long workspaceId) {
```
I checked out the code and this method is used nowhere. Can we delete it?

hreeve-cloudera replied:
Deleted.

biharitomi commented on core/src/main/java/com/sequenceiq/cloudbreak/repository/StackRepository.java:
```diff
@@ -59,6 +59,10 @@
             + "WHERE s.environmentCrn= :environmentCrn AND s.type = :type AND s.terminated=null")
     List<StackStatusView> findByEnvironmentCrnAndStackType(@Param("environmentCrn") String environmentCrn, @Param("type") StackType type);
 
+    @Query("SELECT s FROM Stack s WHERE s.environmentCrn= :environmentCrn AND s.type = :type AND s.workspace.id= :workspaceId AND s.terminated=null")
```
I checked the code locally and the StackService method that using this repo call is used nowhere. Can we delete this repo call please?
I think it could be expensive, we should avoid fetching entire stacks from the database where it is possible.

biharitomi commented on core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:
```diff
@@ -247,11 +252,17 @@ private boolean isEndpointGatewayEnabled(EnvironmentNetworkResponse network) {
         return Optional.ofNullable(knoxTargetGroup);
     }
 
+    private void addTargetToInstanceGroup(InstanceGroup instanceGroup, TargetGroup targetGroup) {
```
This is called from a cycle I would think about having one proper named query to fetch the necessary lines from the two affected tables.

hreeve-cloudera replied:
I'm not sure what you mean. Can you elaborate?

hreeve-cloudera replied:
I've changed the logic here so we're no longer fetching from the database, and have also removed the Stack database fetches from the various handles. The exception is the `CreateLoadBalancerEntityHandler`, because the purpose of that handler is to fetch the database entities and update them. But after that fetch, the stack is passed between handlers instead of making another database call.

biharitomi commented on core/src/main/java/com/sequenceiq/cloudbreak/service/publicendpoint/ClusterPublicEndpointManagementService.java:
```diff
@@ -33,6 +33,13 @@ public boolean provision(Stack stack) {
+    public boolean provisionLoadBalancer(Stack stack) {
+        if (!gatewayPublicEndpointManagementService.renewCertificate(stack)) {
+            LOGGER.warn("Certificate was not updated with load balancer SAN in PEM service.");
+        }
```
can we have a log about the happy path?

biharitomi commented on core/src/main/java/com/sequenceiq/cloudbreak/service/publicendpoint/GatewayPublicEndpointManagementService.java:
```diff
@@ -119,39 +117,47 @@ public String updateDnsEntry(Stack stack, String gatewayIp) {
-                LOGGER.warn("Could not find IP or cloud DNS info for load balancer with endpoint {} ." +
-                    "DNS registration will be skipped.", loadBalancer.getEndpoint());
-            }
+                LOGGER.info("Updating load balancer DNS entries");
```
Can we have a separate method for this else clause?

biharitomi commented on core/src/main/java/com/sequenceiq/cloudbreak/service/publicendpoint/GatewayPublicEndpointManagementService.java:
```diff
@@ -119,39 +117,47 @@ public String updateDnsEntry(Stack stack, String gatewayIp) {
-                dnsRegistered = getDnsManagementService().createOrUpdateDnsEntryWithIp(userCrn, accountId, endpoint,
-                    environment.getName(), false, List.of(loadBalancer.getIp()));
+            if (loadBalancerOptional.isEmpty()) {
+                LOGGER.error("Unable find appropriate load balancer in stack. Load balancer public domain name will not be registered.");
```
WARN level should be enough here.

biharitomi reviewed (changes requested):
Please check my comments!
Mostly about log levels, and one bigger concern about the flow that has two paths.

On the other hand the unit test coverage is not really satisfying on this huge PR, for example for the following classes with modified or new logic that quiet complex there is no new tests to cover them:
```
AwsLoadBalancerLaunchService.java
AwsModelService.java
ClouderaManagerConfigService.java
GatewayPublicEndpointManagementService.java

hreeve-cloudera commented on cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLoadBalancerLaunchService.java:
```diff
@@ -0,0 +1,386 @@
+        return updateCloudformationWithLoadBalancers(ac, stack, resourceNotifier, null);
+    }
+
+    @VisibleForTesting
```
Oops, copied that over from when it wasn't public. :)

hreeve-cloudera commented on cloud-aws/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchServiceLoadBalancerTest.java:
```diff
@@ -60,19 +70,14 @@
 import com.sequenceiq.common.api.type.LoadBalancerType;
 
-@RunWith(MockitoJUnitRunner.class)
+@PrepareForTest(AwsPageCollector.class)
```
That's a question for @afarsang. He wrote that class, and I just followed the pattern that was there. (It was a pain to get it working in the unit test though.

hreeve-cloudera commented on core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/update/loadbalancer/StackLoadBalancerUpdateFlowConfig.java:
```diff
@@ -0,0 +1,139 @@
+            .failureEvent(RESTART_CM_FAILED_EVENT)
+
+        // Path 2: Does not find any load balancers that need to be created, ends flow early
+        .from(CREATING_LOAD_BALANCER_ENTITY_STATE)
```
Thanks for letting me know about that logic. I've added a `StackLoadBalancerUpdateTriggerCondition` class and simplified this flow.

hreeve-cloudera commented on core/src/main/java/com/sequenceiq/cloudbreak/reactor/api/event/stack/loadbalancer/handler/CreateCloudLoadBalancersHandler.java:
```diff
@@ -0,0 +1,95 @@
+                throw new CloudbreakException("Creation failed for load balancers: " + names);
+            }
+
+            LOGGER.debug("Cloud load balancer creation successful");
```
Can you explain to me how the logging levels in CB should be set? When I first started working in the code I tried to be smart about which messages were INFO and which were DEBUG, but the review comments I got back always said to make the messages DEBUG level. I'm making all the level changes you requested, but for future reference.

hreeve-cloudera commented on core/src/main/java/com/sequenceiq/cloudbreak/reactor/api/event/stack/loadbalancer/handler/CreateLoadBalancerEntityHandler.java:
```diff
@@ -0,0 +1,153 @@
+            LOGGER.debug("Creating load balancer entity objects for stack {}", stack.getId());
+            DetailedEnvironmentResponse environment = environmentClientService.getByCrn(stack.getEnvironmentCrn());
+
+            if (environment != null && environment.getNetwork() != null &&
```
This particular check is needed for the `enableEndpointGateway()` method below, but I have added the flow condition and removed some checks from this class.

hreeve-cloudera commented:
> Please check my comments!
> Mostly about log levels, and one bigger concern about the flow that has two paths.
> 
> On the other hand the unit test coverage is not really satisfying on this huge PR, for example for the following classes with modified or new logic that quiet complex there is no new tests to cover them:
> 
> ```
> AwsLoadBalancerLaunchService.java
> AwsModelService.java
> ClouderaManagerConfigService.java
> GatewayPublicEndpointManagementService.java
> ```

`AwsLoadBalancerLaunchService` is tested by `AwsLaunchServiceLoadBalancerTest`. (The names got a little mixed up.) That's not a new test class, but one I wrote to test the AWS load balancer logic earlier. I made `AwsLoadBalancerLaunchService` by moving the existing LB logic from `AwsLaunchService` to its own class. Since the bulk of the logic didn't change I simply updated `AwsLaunchServiceLoadBalancerTest` to use `AwsLoadBalancerLaunchService` instead of `AwsLaunchService`, and added some new test cases for the logic that did change.

`AwsModelService` is again just moving logic that already existed. That class is tested indirectly through `AwsLaunchTest`. Since no new logic was added, I didn't write any new test cases. But if there's a missing test case you think was introduced, let me know.

Added new test cases to `ClouderaManagerConfigServiceTest`.

Added a new test case to `GatewayPublicEndpointManagementService`.

Generally, since a lot of this change was refactoring existing code to new classes and not necessarily writing new logic, existing unit tests are generally still applicable. I did try to write new tests for any new logic I added.

biharitomi reviewed (approved):
Thanks LGTM! I also tested it locally without PEM integration it worked as expected.

sodre90 commented on core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/common/Status.java:
```diff
@@ -50,7 +50,11 @@
-    EXTERNAL_DATABASE_STOP_FAILED(StatusKind.FINAL);
+    EXTERNAL_DATABASE_STOP_FAILED(StatusKind.FINAL),
+    LOAD_BALANCER_UPDATE_IN_PROGRESS(StatusKind.PROGRESS),
+    LOAD_BALANCER_UPDATE_FINISHED(StatusKind.FINAL),
```
In DetailedStackStatus.java:
LOAD_BALANCER_UPDATE_FINISHED(Status.AVAILABLE)

It is AVAILABLE not LOAD_BALANCER_UPDATE_FINISHED and it is correct. I think we should remove LOAD_BALANCER_UPDATE_FINISHED.

hreeve-cloudera replied:
Hi @sodre90. This PR was closed 2.5 months ago. I get your point, but it's a little late to make changes to it now. Can I ask what prompted this?
